pkg/envinfo/url.go

```diff
@@ -0,0 +1,275 @@
+package envinfo
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	devfilev1 "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/library/pkg/devfile/generator"
+	"github.com/devfile/library/pkg/devfile/parser"
+	"github.com/openshift/odo/pkg/localConfigProvider"
+	"github.com/openshift/odo/pkg/odo/util/validation"
+	"github.com/openshift/odo/pkg/util"
+	"github.com/pkg/errors"
+)
+
+// GetPorts returns the ports, returns default if nil
+func (ei *EnvInfo) GetPorts() []string {
+
+	containerComponents := ei.devfileObj.Data.GetDevfileContainerComponents()
+	portMap := make(map[string]bool)
+
+	var portList []string
+	for _, component := range containerComponents {
+		for _, endpoint := range component.Container.Endpoints {
+			portMap[strconv.FormatInt(int64(endpoint.TargetPort), 10)] = true
+		}
+	}
+
+	for port := range portMap {
+		portList = append(portList, port)
+	}
+
+	sort.Strings(portList)
+	return portList
+}
+
+// CompleteURL completes the given URL with default values
+func (ei *EnvInfo) CompleteURL(url *localConfigProvider.LocalURL) error {
+	if len(url.Path) > 0 && (strings.HasPrefix(url.Path, "/") || strings.HasPrefix(url.Path, "\\")) {
+		if len(url.Path) <= 1 {
+			url.Path = ""
+		} else {
+			// remove the leading / or \ from provided path
+			url.Path = string([]rune(url.Path)[1:])
+		}
+	}
+	// add leading / to path, if the path provided is empty, it will be set to / which is the default value of path
+	url.Path = "/" + url.Path
+
+	containerComponents := ei.devfileObj.Data.GetDevfileContainerComponents()
+	if len(containerComponents) == 0 {
+		return nil
+	}
+
+	if len(url.Container) > 0 {
```

Comment:
This check seems out of place to me. We're setting the variable `containerComponents` in an earlier step. But we're working on `url.Container` here. What's the significance?

Reply:
Fixed and added comments